flow/project.py

```diff
@@ -133,52 +133,64 @@ def not_(cls, condition):
         return cls(lambda job: not condition(job))
 
 
-class _pre(_condition):
+def setup_pre_conditions_class(cls):
 
-    def __call__(self, func):
-        pre_conditions = getattr(func, '_flow_pre', list())
-        pre_conditions.insert(0, self.condition)
-        func._flow_pre = pre_conditions
-        return func
+    class _pre(_condition):
 
-    @classmethod
-    def copy_from(cls, *other_funcs):
-        "True if and only if all pre conditions of other function(s) are met."
-        def metacondition(job):
-            return all(c(job)
-                       for other_func in other_funcs
-                       for c in getattr(other_func, '_flow_pre', list()))
-        return cls(metacondition)
+        owner_class = cls
 
-    @classmethod
-    def after(cls, *other_funcs):
-        "True if and only if all post conditions of other function(s) are met."
-        def metacondition(job):
-            return all(c(job)
-                       for other_func in other_funcs
-                       for c in getattr(other_func, '_flow_post', list()))
-        return cls(metacondition)
+        def __init__(self, condition):
+            self.condition = condition
 
+        def __call__(self, func):
+            self.owner_class._OPERATION_PRE_CONDITIONS[func].insert(0, self.condition)
```

Comment:
Out of curiosity why do we use insert rather than append?

Reply:
The nature of decorators as *functions of functions* means that the ones defined on top are executed last. However, to make it easier for users, we have specified that decorators that are _defined_ earlier have higher priority. That's why we insert them at the top of the list.

Reply:
See also here: http://docs.signac.io/en/latest/flow-project.html#defining-a-workflow